site-src/geps/gep-1424.md

```diff
@@ -0,0 +1,384 @@
+
+### Omitting `backendRefs`
+
+Implementations SHOULD support a terse syntax which allows omitting `backendRefs` to avoid unnecessary redundancy for simple configurations. If no `backendRefs` are specified, implementations should direct traffic to the default endpoints of the `parentRef` `Service`. If the `parentRef` does not have any endpoints (e.g. if it is a `Service` without selectors), implementations MUST return an HTTP 503 response code (see discussion in [#1210](https://github.com/kubernetes-sigs/gateway-api/pull/1210)).
+
+This syntax has limited utility currently, but could become more relevant if additional functionality beyond traffic splitting is added to the `HTTPRoute` resource. `HTTPRoutes` configured in this way MAY be used to manually enroll services into a mesh (which could trigger behavior like injecting a sidecar proxy) if they have not already been enrolled by some other mechanism (as proposed in [GEP-1291: Mesh Representation](https://docs.google.com/document/d/1oyA9uUH7pNNxxwy3WZGSWx-edHDBLrujcezr8q3el70/edit) or similar).
+
+```
+metadata:
+  name: foo_route
+  namespace: store
+spec:
+  parentRefs:
+  - kind: Service
+    name: foo
+  rules:
+    matches:
+    - path:
+        value: "/bar"
+```
+
+The example above would drop all incoming traffic for HTTP paths other than `/bar` to the `foo` `Service`, following the existing spec to return HTTP 404 response codes for unmatched requests, and HTTP 500 response codes for requests excluded due to an [`HTTPRouteFilter`](https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io/v1beta1.HTTPRouteFilter).
+
+### Allowed `Service` types
+
+Valid `Service` resources to be selected as a `parentRef` SHOULD provision one or more virtual IP(s) and DNS hostname(s) within the Kubernetes cluster, and include the default [`ClusterIP`](https://kubernetes.io/docs/concepts/services-networking/service/#publishing-services-service-types) `Service` type, with or without selectors.
+
+`Service` resource with [`type: NodePort`](https://kubernetes.io/docs/concepts/services-networking/service/#type-nodeport) or [`type: LoadBalancer`](https://kubernetes.io/docs/concepts/services-networking/service/#loadbalancer) MUST NOT be allowed as `parentRefs` or `backendRefs`. It should isntead be preferred to expose services publicly through the North/South Gateway API interfaces.
```

Comment:
```suggestion
`Service` resource with [`type: NodePort`](https://kubernetes.io/docs/concepts/services-networking/service/#type-nodeport) or [`type: LoadBalancer`](https://kubernetes.io/docs/concepts/services-networking/service/#loadbalancer) MAY be allowed as `parentRefs` or `backendRefs`, as these do provision virtual IPs and are effectively `ClusterIP` services with additional functionality, but it should typically be preferred to expose services publicly through the North/South Gateway API interfaces instead.
```

Reply:
I would vote for being restrictive at first and opening things up as we find use cases for them -- my gut feel is that I'd much rather handle ingress with an ingress controller.

Reply:
I would allow them. These are supersets of ClusterIP types.

Reply:
For a bit more context, I raised some concerns here because it feels somewhat arbitrary to not support NodePort or LoadBalancer Services. LoadBalancer Services are a superset of NodePort Services which are a superset of ClusterIP Services. This means that we're effectively saying that we don't support a set of Services that happen to have some extra functionality on top. I agree with Mike's suggested change above. Essentially I think our stance should be that:

1. ClusterIP Services are recommended
2. The ClusterIP portion of NodePort and LoadBalancer Services are supported
3. We recommend using ClusterIP Services over alternative types because the additional functionality added by NodePort and LoadBalancer Services will be ignored by mesh implementations.

Reply:
I'm going to merge this with the proposed language (`MAY` and the recommendation to prefer `ClusterIP` services). For a provisional GEP that feels like sufficient leeway to allow implementations to support them, but clear that it's not required. We should definitely start with conformance tests only for ClusterIP services, and use the existing "opt-in" functionality for `NodePort` or `LoadBalancer` `Service` types if/when those tests are added.